### PR TITLE
Image Array

### PR DESCRIPTION
**Pull request by changhiskhan**
Make it easier to convert from iterable of Image and chunked arrays

### CODE REVIEW

File: python/lance/types/image.py [hunk 1]
```diff
@@ -218,3 +263,70 @@ class ImageUriScalar(pa.ExtensionScalar):
...
+                return ImageArray.from_pandas(storage)
+
+        return pa.ExtensionArray.from_pandas(
+            obj, mask=mask, type=type, safe=safe, memory_pool=memory_pool
+        )
+
+    @staticmethod
+    def from_images(images, type=None, mask=None, safe=True, memory_pool=None):
+        """
+        Create an ImageUriArray from Image instances
```
Comment (eddyxu) [inline python/lance/types/image.py:301]: Does this only create URI array?

File: python/lance/types/image.py [hunk 2]
```diff
@@ -218,3 +263,70 @@ class ImageUriScalar(pa.ExtensionScalar):
...
+        Parameters
+        ----------
+        images : sequence of Image
+        type : DataType, default None
+            If not specified then use the arrow_dtype of the first Image
+            instance
+        mask : array[bool], optional
```
Comment (eddyxu) [inline python/lance/types/image.py:308]: Can this just be the None values in the images list.

File: python/lance/types/image.py [hunk 3]
```diff
@@ -218,3 +263,70 @@ class ImageUriScalar(pa.ExtensionScalar):
 
     def as_py(self) -> Image:
         return ImageUri(self.value.as_py())
+
+
+class ImageArray(pa.ExtensionArray):
+    @staticmethod
+    def from_pandas(obj, mask=None, type=None, safe=True, memory_pool=None):
```
Comment (eddyxu) [inline python/lance/types/image.py:270]: Add docstring?

File: python/lance/types/image.py [hunk 4]
```diff
@@ -218,3 +263,70 @@ class ImageUriScalar(pa.ExtensionScalar):
...
+            obj, mask=mask, type=type, safe=safe, memory_pool=memory_pool
+        )
+
+    @staticmethod
+    def from_images(images, type=None, mask=None, safe=True, memory_pool=None):
+        """
+        Create an ImageUriArray from Image instances
+        Parameters
+        ----------
+        images : sequence of Image
```
Comment (eddyxu) [inline python/lance/types/image.py:304]: does it expect list / set / tuple or pd.Series / np.array?

File: python/lance/types/image.py [hunk 5]
```diff
@@ -218,3 +263,70 @@ class ImageUriScalar(pa.ExtensionScalar):
...
+        if len(images) > 0:
+            type = images[0].arrow_dtype
+            storage = pa.array(
+                [im.to_arrow_storage() for im in images],
+                mask=mask,
+                safe=safe,
+                memory_pool=memory_pool,
+            )
+        else:
+            storage = pa.array(images, type=type)
```
Comment (eddyxu) [inline python/lance/types/image.py:325]: If `images` is a sequence, this else branch means the images list is empty?

Btw, it seems the type parameter is only used here.

File: python/lance/tests/test_types.py [hunk 6]
```diff
@@ -147,3 +176,7 @@ def test_pickle(tmp_path):
         pickle.dump(img, fh)
     with (tmp_path / "image").open("rb") as fh:
         assert img == pickle.load(fh)
+
+
+def test_image_array(tmp_path):
```
[the remark this answers is not in the document]
Comment (changhiskhan) [reply python/lance/tests/test_types.py:181]: fixed

File: python/lance/types/image.py (hunk 1 above)
Comment (changhiskhan) [reply python/lance/types/image.py:301]: copy-paste error. fixed.

File: python/lance/types/image.py (hunk 3 above)
Comment (changhiskhan) [reply python/lance/types/image.py:270]: added.

File: python/lance/types/image.py (hunk 4 above)
Comment (changhiskhan) [reply python/lance/types/image.py:304]: anything that's list-like. I can change this to "sequence / iterable / ndarray / pandas.Series of Image" to be more explicit (matches pa.array docstring)

File: python/lance/types/image.py (hunk 5 above)
Comment (changhiskhan) [reply python/lance/types/image.py:325]: i'll be more explicit and change this to pa.array([], type=type).

If the input is non-empty, then the Image instance should determine the type and not the `type` argument. wdyt?

**Review by eddyxu — APPROVED** [review]

**Pull request merged**